r/src/array_from_vector.cpp

```diff
@@ -797,79 +923,112 @@ Status GetConverter(const std::shared_ptr<DataType>& type,
   return Status::NotImplemented("type not implemented");
 }
 
-template <typename Type>
-std::shared_ptr<arrow::DataType> GetFactorTypeImpl(bool ordered) {
-  return dictionary(std::make_shared<Type>(), arrow::utf8(), ordered);
+static inline std::shared_ptr<arrow::DataType> IndexTypeForFactors(int n_factors) {
+  if (n_factors < INT8_MAX) {
+    return arrow::int8();
```

Comment:
It's still returning a signed integer (from the C api). Sticking to signed will provide better experience with other languages (JVM) and Parquet since they mostly used signed ints.